src/metabase/util/queue.clj

```diff
@@ -63,24 +70,141 @@
   (DelayQueue.))
 
 (defn put-with-delay!
-  "Put an item on the delay queue, with a delay given in milliseconds."
+  "Put an item on a delay queue, with a delay given in milliseconds."
   [^DelayQueue queue delay-ms value]
   (.offer queue (->DelayValue value (.plus (Instant/now) (Duration/ofMillis delay-ms)))))
 
-(defn- take-delayed-batch* [^DelayQueue queue max-items ^long poll-ms acc]
+(defn- take-batch* [^BlockingQueue queue ^long max-messages ^long latest-time acc]
   (loop [acc acc]
-    (if (>= (count acc) max-items)
-      acc
-      (if-let [item (if (pos? poll-ms)
-                      (.poll queue poll-ms TimeUnit/MILLISECONDS)
-                      (.poll queue))]
-        (recur (conj acc (:value item)))
-        (not-empty acc)))))
-
-(defn take-delayed-batch!
-  "Get up to `max-items` of the ready items off a given delay queue."
-  ([queue max-items]
-   (take-delayed-batch* queue max-items 0 []))
-  ([^DelayQueue queue max-items ^long max-first-ms ^long max-next-ms]
-   (when-let [fst (.poll queue max-first-ms TimeUnit/MILLISECONDS)]
-     (take-delayed-batch* queue max-items max-next-ms [(:value fst)]))))
+    (let [remaining-ms (- latest-time (System/currentTimeMillis))]
```

Comment:
We need to take some care with changing the wait semantics like this - for large batch sizes it can make a big difference to behavior. I'd prefer we kept this `debounce-ms` parameter working as is, and added a new `max-ms` parameter if you think it's worthwhile.
 
For measuring elapsed time we should always use `System/nanoTime`, `currentTimeMillis` is only meant for wall clock time. Life is simpler without a system call, leap seconds, etc.

Reply:
Went back to the max-next-ms style logic